### PR TITLE
ISAICP-5791: Don't allow abuse of password recovery

### PR DESCRIPTION
**Pull request by claudiu-cristea**

### CODE REVIEW

File: web/modules/custom/eulogin/src/Event/Subscriber/JoinupEuLoginKernelSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+      'system.csrftoken',
+      // Still able to contact support.
+      'contact_form.contact_page',
+      // Can see their account page.
+      'user.page',
+      // Able to log out.
+      'user.logout',
+      // A Well-Known URL for Changing Passwords.
+      // @see https://wicg.github.io/change-password-url
+      'user.well-known.change_password',
```
Comment (pfrenssen) [inline web/modules/custom/eulogin/src/Event/Subscriber/JoinupEuLoginKernelSubscriber.php]: Interesting :)

**Comment by pfrenssen** [comment]
Looking good, just some nitpicks!

File: web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  /**
+   * Test limited access for one-time-login sessions.
+   */
+  public function testLimitedAccess(): void {
+    $state = \Drupal::state();
+
+    // This test needs the limited access functionality enabled.
+    $state->delete('joinup_eulogin.limited_access');
```
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php]: I prefer to delete the in order to keep the state store clean. This variable makes sense only when disabling the "limit access". I prefer that we keep the state store clean with the normal behaviour and that is "limited access". Or should I revert the variable, i.e. `joinup_eulogin.disable_limited_access`, and this defaults to FALSE? Then it makes sense to delete it with the normal behaviour (limited access).

[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php]: Finally I went with the second option: I've been reverted the logic of the kill-switch.

[the remark this answers is not in the document]
Comment (pfrenssen) [reply web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php]: Perfect solution, thanks!

File: web/modules/custom/joinup_core/tests/src/ExistingSite/JoinupExistingSiteTestBase.php [hunk 3]
```diff
@@ -5,34 +5,66 @@
...
+
     // As ExistingSiteBase tests are running without javascript, we disable
-    // Antibot during the tests run.
-    static::disableAntibot();
+    // Antibot and Honeypot functionality during the tests run.
+    $this->disableAntibot();
+    $this->disableHoneypot();
+
+    // Disable limited access functionality.
+    \Drupal::state()->set('joinup_eulogin.limited_access', FALSE);
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply web/modules/custom/joinup_core/tests/src/ExistingSite/JoinupExistingSiteTestBase.php]: Excellent explanation, our future maintainers are very grateful :smile:

File: web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php [hunk 4]
```diff
@@ -213,4 +216,40 @@ protected function extractPasswordResetUrlFromMail(): string {
...
+    // Borrowed from CharacterTypes::validate().
+    $valid_password = function (string $password): bool {
+      $character_sets = count(array_filter([
+        preg_match('/[a-z]/', $password),
+        preg_match('/[A-Z]/', $password),
+        preg_match('/[0-9]/', $password),
+        preg_match('/[^a-zA-Z0-9]/', $password),
+      ]));
+      return $character_sets >= 3;
+    };
```
Comment (pfrenssen) [inline web/modules/custom/eulogin/tests/src/ExistingSite/JoinupEuLoginTest.php]: :+1:

**Pull request merged**